nomad/structs/service_registration.go

```diff
@@ -105,3 +141,90 @@ func (s *ServiceRegistration) Equals(o *ServiceRegistration) bool {
 	}
 	return true
 }
+
+// Validate ensures the upserted service registration contains valid
+// information and routing capabilities. Objects should never fail here as
+// Nomad controls the entire registration process; but it's possible
+// configuration problems could cause failures.
+func (s *ServiceRegistration) Validate() error {
+	if ipaddr.IsAny(s.Address) {
+		return fmt.Errorf("invalid service registration address")
+	}
+	return nil
+}
+
+// ServiceRegistrationUpsertRequest is the request object used to upsert one or
+// more service registrations.
+type ServiceRegistrationUpsertRequest struct {
+	Services []*ServiceRegistration
+	WriteRequest
+}
+
+// ServiceRegistrationUpsertResponse is the response object when one or more
+// service registrations have been successfully upserted into state.
+type ServiceRegistrationUpsertResponse struct {
+	WriteMeta
+}
+
+// ServiceRegistrationDeleteByIDRequest is the request object to delete a
+// service registration as specified by the ID parameter.
+type ServiceRegistrationDeleteByIDRequest struct {
+	ID string
+	WriteRequest
+}
+
+// ServiceRegistrationDeleteByIDResponse is the response object when performing a
+// deletion of an individual service registration.
+type ServiceRegistrationDeleteByIDResponse struct {
+	WriteMeta
+}
+
+// ServiceRegistrationDeleteByNodeIDRequest is the request object to delete all
+// service registrations assigned to a particular node.
+type ServiceRegistrationDeleteByNodeIDRequest struct {
+	NodeID string
+	WriteRequest
+}
+
+// ServiceRegistrationDeleteByNodeIDResponse is the response object when
+// performing a deletion of all service registrations assigned to a particular
+// node.
+type ServiceRegistrationDeleteByNodeIDResponse struct {
+	WriteMeta
+}
+
+// ServiceRegistrationListRequest is the request object when performing service
+// registration listings.
+type ServiceRegistrationListRequest struct {
+	QueryOptions
+}
+
+// ServiceRegistrationListResponse is the response object when performing a
+// list of services. This is specifically concise to reduce the serialisation
+// and network costs endpoint incur, particularly when performing blocking list
+// queries.
+type ServiceRegistrationListResponse struct {
+	Services []*ServiceRegistrationListStub
+	QueryMeta
+}
+
+// ServiceRegistrationListStub is the object which contains a list of namespace
+// service registrations and their tags.
+type ServiceRegistrationListStub struct {
+	Namespace string
+	Services  map[string][]string
```

Comment:
A map means we can never sort/paginate this and a `[]string` means we can never add more fields to the stub.

This would be more future proof... think it's worth it?
```go
type ServiceRegistrationStub struct {
	ServiceName string
	Tags        []string
}

type ServiceRegistrationListStub struct {
	Namespace string
	Services  []*ServiceRegistrationStub
}
```

I know we don't have solid use cases for this outside of "match Consul" and "peek at services available" which your implementation clearly satisfies... :thinking:

Reply:
Good shout. I really struggled to model this correctly but this approach looks great. I have made the change to this PR already and will update the API and SDK one to reflect this also.